aten/src/ATen/cuda/CUDAFuture.cpp

```diff
@@ -70,25 +117,33 @@ void CUDAFuture::preMarkCompletedHook(
   // Start by performing all steps that can throw, before setting any field.
   std::vector<std::reference_wrapper<const at::DataPtr>> actualDataPtrs =
       dataPtrs.has_value() ? std::move(*dataPtrs) : extractDataPtrs(value);
+  std::vector<c10::DeviceIndex> usedDevices =
+      getDevicesOfDataPtrs(actualDataPtrs);
+  if (devices_.has_value()) {
+    std::vector<c10::DeviceIndex> excessDevices;
+    std::set_difference(
+        usedDevices.begin(),
+        usedDevices.end(),
+        devices_->begin(),
+        devices_->end(),
+        std::back_inserter(excessDevices));
+    TORCH_CHECK_VALUE(
+        excessDevices.empty(),
+        "The result contained tensors residing on device(s) ",
+        formatSetOfDevices(excessDevices),
+        " which are not among the expected device(s) ",
```

Comment:
nit: mention that the user can specify these devices when constructing cuda future?

Reply:
Uhm, that makes sense, but I'm not sure how we can do it properly: this error could also be raised when a callback returns a value on a "bad" device, but the parent future could come from a variety of places (hand-constructed, returned from ProcessGroupNCCL, or from the TensorPipeAgent) each of which has its own way of setting the supported devices, and I don't think we want to list and explain all of them here. Do you have a wording that you think would be generic but still convey the message?

Reply:
I see, its a good point that it can come from a variety of places so probably logging the devices like we do is sufficient here.